R/compute_portfolio_returns.R

```diff
@@ -0,0 +1,304 @@
+#'   \itemize{
+#'     \item `portfolio`: The portfolio identifier.
+#'     \item `date`: The date of the portfolio return.
+#'     \item `ret_excess_vw`: The value-weighted excess return of the portfolio.
+#'     \item `ret_excess_ew`: The equal-weighted excess return of the portfolio.
+#'   }
+#'
+#' @note
+#' Ensure that the `sorting_data` contains all the required columns: the specified sorting variables,
+#' `mktcap_lag`, and `ret_excess`. The function will stop and throw an error if any required columns are missing.
+#'
+#' @export
+#'
+#' @examples
+#' # Univariate sorting with monthly rebalancing
+#' data <- data.frame(
+#'   permno = 1:100,
+#'   date = rep(seq.Date(from = as.Date("2020-01-01"), by = "month", length.out = 100), each = 10),
+#'   mktcap_lag = runif(100, 100, 1000),
+#'   ret_excess = rnorm(100),
+#'   size = runif(100, 50, 150)
+#' )
+#' compute_portfolio_returns(data, sorting_variables = "size", sorting_method = "univariate", n_portfolios = 5)
+#'
+#' # Bivariate dependent sorting with annual rebalancing
+#' compute_portfolio_returns(data, sorting_variables = c("size", "mktcap_lag"), sorting_method = "bivariate-dependent", n_portfolios = c(3, 3), rebalancing_frequency = "annual")
+#'
+compute_portfolio_returns <- function(
+    sorting_data,
+    sorting_variables,
+    sorting_method,
+    rebalancing_frequency = "monthly",
+    n_portfolios = NULL, # must be a vector
+    percentiles = NULL, # must be a list with length == length(sorting_variables)
+    breakpoint_exchanges = NULL
+) {
+
+  if (is.null(sorting_variables) || length(sorting_variables) == 0) {
+    cli::cli_abort("You must provide at least one sorting variable.")
+  }
+
+  if (!sorting_method %in% c("univariate", "bivariate-dependent", "bivariate-independent")) {
+    cli::cli_abort("Invalid sorting method. Choose 'univariate', 'bivariate-dependent', or 'bivariate-independent'.")
+  }
+
+  if (is.null(n_portfolios) && is.null(percentiles)) {
+    cli::cli_abort("You must provide either 'n_portfolios' or 'percentiles'.")
+  }
+
+  if (!is.null(n_portfolios) && !is.null(percentiles)) {
+    cli::cli_abort("Please provide either 'n_portfolios' or 'percentiles', not both.")
+  }
+
+  if (!is.null(n_portfolios) && length(n_portfolios) != length(sorting_variables)) {
+    cli::cli_abort("'n_portfolios' must have the same length as 'sorting_variables'.")
+  }
+
+  if (!is.null(percentiles)) {
+    if (!is.list(percentiles) || length(percentiles) != length(sorting_variables)) {
+      cli::cli_abort("'percentiles' must be a list with the same length as 'sorting_variables'.")
+    }
+  }
+
+  if (!is.null(breakpoint_exchanges)) {
+    if (!("exchange" %in% colnames(sorting_data))) {
+      cli::cli_abort("The 'sorting_data' must contain an 'exchange' column when 'breakpoint_exchanges' is provided.")
+    }
+  }
+
+  required_columns <- c(sorting_variables, "mktcap_lag", "ret_excess")
```

Comment:
I don't see why mktcap_lag should be enforced if I only care about equal weighted portfolios. Can't this be a soft requirement in the sense that value-weighted returns are omitted if mktcap_lag is not provided. Or we use the aggregation method as a function input (value-weighted or equal-weighted).

Reply:
Fine for me to throw a message and omit value-weighted returns if mktcap_lag is missing. I'm against using aggregation method as function input because people have to run it twice to get these outputs (was like this in original implementation).